lib/cdo/contact_rollups.rb

```diff
@@ -326,6 +327,40 @@ def self.insert_from_dashboard_census_submissions
     log_completion(start)
   end
 
+  def self.update_geo_from_school_data
+    start = Time.now
+    log "Updating user geo data from school data"
+
+    # State for schools is stored in state abbreviation. We need to convert
+    # to state name, so do this row-by-row using existing Ruby code for that
+    # conversion.
+
+    sql = "
+    SELECT users.email, schools.city, schools.state, schools.zip
+    FROM users
+    INNER JOIN school_infos ON school_infos.id = users.school_info_id
+    INNER JOIN schools ON schools.id = school_infos.school_id"
+
+    dataset = DASHBOARD_REPORTING_DB_READER[sql]
+
+    dataset.each do |user_and_geo|
+      state_code = user_and_geo[:state]
+      # convert from state code to state name
+      state = get_us_state_from_abbr(state_code, true)
+      next unless state.presence
+      city = user_and_geo[:city]
+      zip = user_and_geo[:zip]
+      email = user_and_geo[:email]
+      # update the user's city/state/zip
+      PEGASUS_REPORTING_DB_WRITER[DEST_TABLE_NAME.to_sym].where(email: email).
```

Comment:
How come we don't use this nice active record logic elsewhere in the code? This is so much cleaner than the `ON DUPLICATE KEY UPDATE...` type stuff I imitated in previous changes.

Reply:
Asking for my edification more than anything.

Reply:
Because this is way slower! This executes row by row. We need to do it this way in this case because we have to translate state abbreviation to state name. We had some Ruby code that does it, there are currently not that many rows with this data (~100K), so doing it this way here doesn't explode the process time. We use the "bulk SQL" approach in general because a lot of the data is big (25M rows). The rollup process would take more than 24 hours if we tried to do all the steps on a row-by-row basis.